### PR TITLE
Fix httpjson split issue on empty root array

### LINKED ISSUE #28221: Filebeat httpjson cursor not saved when splitting on empty array.
On filebeat 7.14.1, but probably other recent versions, I'm using httpjson to query an API for Proofpoint.

The Proofpoint SIEM API is documented here: https://help.proofpoint.com/Threat_Insight_Dashboard/API_Documentation/SIEM_API

A sample payload is here: https://help.proofpoint.com/Threat_Insight_Dashboard/API_Documentation/SIEM_API#section_29

Below is an example adapted for just `messagesBlocked` per the /v2/siem/messages/blocked endpoint.
```
{
   "queryEndTime":"2016-06-24T21:36:00Z",
   "messagesBlocked":[
      {
         "GUID":"c26dbea0-80d5-463b-b93c-4e8b708219ce",
         "QID":"r2FNwRHF004109",
         "ccAddresses":[
            "bruce.wayne@university-of-education.zz"
         ],
         "clusterId":"pharmtech_hosted",
         "completelyRewritten":"true",
         "fromAddress":"badguy@evil.zz",
         "headerCC":"\"Bruce Wayne\" <bruce.wayne@university-of-education.zz>",
         "headerFrom":"\"A. Badguy\" <badguy@evil.zz>",
         "headerReplyTo":null,
         "headerTo":"\"Clark Kent\" <clark.kent@pharmtech.zz>; \"Diana Prince\" <diana.prince@pharmtech.zz>",
         "impostorScore":0,
         "malwareScore":100,
         "messageID":"20160624211145.62086.mail@evil.zz",
         "messageParts":[
            {
               "contentType":"text/plain",
               "disposition":"inline",
               "filename":"text.txt",
               "md5":"008c5926ca861023c1d2a36653fd88e2",
               "oContentType":"text/plain",
               "sandboxStatus":"unsupported",
               "sha256":"85738f8f9a7f1b04b5329c590ebcb9e425925c6d0984089c43a022de4f19c281"
            },
            {
               "contentType":"application/pdf",
               "disposition":"attached",
               "filename":"Invoice for Pharmtech.pdf",
               "md5":"5873c7d37608e0d49bcaa6f32b6c731f",
               "oContentType":"application/pdf",
               "sandboxStatus":"threat",
               "sha256":"2fab740f143fc1aa4c1cd0146d334c5593b1428f6d062b2c406e5efe8abe95ca"
            }
         ],
         "messageTime":"2016-06-24T21:18:38.000Z",
         "modulesRun":[
            "pdr",
            "sandbox",
            "spam",
            "urldefense"
         ],
         "phishScore":46,
         "policyRoutes":[
            "default_inbound",
            "executives"
         ],
         "quarantineFolder":"Attachment Defense",
         "quarantineRule":"module.sandbox.threat",
         "recipient":[
            "clark.kent@pharmtech.zz",
            "diana.prince@pharmtech.zz"
         ],
         "replyToAddress":null,
         "sender":"e99d7ed5580193f36a51f597bc2c0210@evil.zz",
         "senderIP":"192.0.2.255",
         "spamScore":4,
         "subject":"Please find a totally safe invoice attached.",
         "threatsInfoMap":[
            {
               "campaignId":"46e01b8a-c899-404d-bcd9-189bb393d1a7",
               "classification":"MALWARE",
               "threat":"2fab740f143fc1aa4c1cd0146d334c5593b1428f6d062b2c406e5efe8abe95ca",
               "threatId":"2fab740f143fc1aa4c1cd0146d334c5593b1428f6d062b2c406e5efe8abe95ca",
               "threatStatus":"active",
               "threatTime":"2016-06-24T21:18:38.000Z",
               "threatType":"ATTACHMENT",
               "threatUrl":"https://threatinsight.proofpoint.com/#/73aa0499-dfc8-75eb-1de8-a471b24a2e75/threat/u/2fab740f143fc1aa4c1cd0146d334c5593b1428f6d062b2c406e5efe8abe95ca"
            },
            {
               "campaignId":"46e01b8a-c899-404d-bcd9-189bb393d1a7",
               "classification":"MALWARE",
               "threat":"badsite.zz",
               "threatId":"3ba97fc852c66a7ba761450edfdfb9f4ffab74715b591294f78b5e37a76481aa",
               "threatTime":"2016-06-24T21:18:07.000Z",
               "threatType":"URL",
               "threatUrl":"https://threatinsight.proofpoint.com/#/73aa0499-dfc8-75eb-1de8-a471b24a2e75/threat/u/3ba97fc852c66a7ba761450edfdfb9f4ffab74715b591294f78b5e37a76481aa"
            }
         ],
         "toAddresses":[
            "clark.kent@pharmtech.zz",
            "diana.prince@pharmtech.zz"
         ]
      },
      "xmailer":"Spambot v2.5"
   ]
}

```

I'm using  `response.split` like this:
```
  response.split:
     target: body.messagesBlocked
```

This works great, except when the payload has no messages and looks like this:
```
{
   "queryEndTime":"2016-06-24T21:36:00Z",
   "messagesBlocked":[]
}
```

The problem is that this is valid.  Of course I may want to drop these events later, but this failure causes an `the requested root field is empty` debug message.  This prevents the cursor from being updated with the `queryEndTime`.

My suggestion is to return an empty event when splitting an empty array so scenarios like this aren't missed.

My proposal is to update [L138](https://github.com/elastic/beats/blob/cca59bad174b56ced2f74bccc6694f233752eecd/x-pack/filebeat/input/httpjson/internal/v2/split.go#L138) to look like this:
```
			if s.isRoot {
				ch <- maybeMsg{msg: common.MapStr{}}
				return errEmptyRootField
			}
```

This causes the event to fire and the cursor to behave as expected.  Personally I am okay if the cursor can still be updated and no event resulting from the split, but I'm not sure if that complicates things.

I'm open to discuss this, especially if there may be a smarter approach here.  I understand that an empty event resulting from the split could break following logic that depends on fields existing.  

Maybe if there was a way for the cursor to be optionally evaluated before the split then this could handle a wider range of scenarios?

Thanks!

### PR DESCRIPTION
**Pull request by kcreddy**
## What does this PR do?
Fix httpjson split issue on empty root array. This fix publishes the event with key in case of empty root array

## Why is it important?
This is a bug fix on filebeat httpjson input where no event is being published when splitting empty root array. The desired outcome is to publish event with array.

## Checklist

- [x] My code follows the style guidelines of this project
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally
```
cd x-pack/filebeat/input/httpjson; go test -v -run TestSplit

=== RUN   TestSplit
=== RUN   TestSplit/Two_nested_Split_Arrays_with_keep_parent
=== RUN   TestSplit/A_nested_array_with_a_nested_map
=== RUN   TestSplit/A_nested_array_with_a_nested_map_with_transforms
=== RUN   TestSplit/A_nested_array_with_a_nested_array_in_an_object
=== RUN   TestSplit/A_nested_array_with_an_empty_nested_array_in_an_object_publishes_without_the_key
=== RUN   TestSplit/First_level_split_publishes_with_key_if_no_events
=== RUN   TestSplit/Changes_must_be_local_to_parent_when_nested_splits
=== RUN   TestSplit/Split_string
=== RUN   TestSplit/An_empty_array_in_an_object
=== RUN   TestSplit/A_missing_array_in_an_object
=== RUN   TestSplit/An_empty_map_in_an_object
=== RUN   TestSplit/A_missing_map_in_an_object
=== RUN   TestSplit/An_empty_string
=== RUN   TestSplit/A_missing_string
--- PASS: TestSplit (0.00s)
    --- PASS: TestSplit/Two_nested_Split_Arrays_with_keep_parent (0.00s)
    --- PASS: TestSplit/A_nested_array_with_a_nested_map (0.00s)
    --- PASS: TestSplit/A_nested_array_with_a_nested_map_with_transforms (0.00s)
    --- PASS: TestSplit/A_nested_array_with_a_nested_array_in_an_object (0.00s)
    --- PASS: TestSplit/A_nested_array_with_an_empty_nested_array_in_an_object_publishes_without_the_key (0.00s)
    --- PASS: TestSplit/First_level_split_publishes_with_key_if_no_events (0.00s)
    --- PASS: TestSplit/Changes_must_be_local_to_parent_when_nested_splits (0.00s)
    --- PASS: TestSplit/Split_string (0.00s)
    --- PASS: TestSplit/An_empty_array_in_an_object (0.00s)
    --- PASS: TestSplit/A_missing_array_in_an_object (0.00s)
    --- PASS: TestSplit/An_empty_map_in_an_object (0.00s)
    --- PASS: TestSplit/A_missing_map_in_an_object (0.00s)
    --- PASS: TestSplit/An_empty_string (0.00s)
    --- PASS: TestSplit/A_missing_string (0.00s)
PASS
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/httpjson      0.289s

```

## Related issues
- Closes #28221

### CODE REVIEW

**Review by efd6 — COMMENTED** [review]
This should have a line in CHANGELOG.next.asciidoc and probably wants a backport to the relevant versions.

**Comment by P1llus** [comment]
@marc-gr does this only need to be applied for split on Array, or also on map/string?

**Comment by ShourieG** [comment]
@kcreddy , please resolve the conflict , otherwise LGTM

File: CHANGELOG.next.asciidoc
Comment (efd6) [inline CHANGELOG.next.asciidoc:58]:
```suggestion
- Fix handling of empty array in httpjson input. {pull}32001[32001]
```

**Comment by andrewkroh** [comment]
Other than the changelog conflicts, does this need anything else before merging? @kcreddy

**Comment by kcreddy** [comment]
Sorry my bad, I thought this was merged long back. Resolved merge conflicts. This can be merged once approved

File: CHANGELOG.next.asciidoc
Comment (kcreddy) [reply CHANGELOG.next.asciidoc]: Added the line at the end

**Review by ShourieG — APPROVED** [review]
LGTM

File: CHANGELOG.next.asciidoc
Comment (efd6) [inline CHANGELOG.next.asciidoc:88]: I think these should be removed.

**Review by efd6 — COMMENTED** [review]
LGTM after changelog is fixed.

File: CHANGELOG.next.asciidoc
Comment (kcreddy) [reply CHANGELOG.next.asciidoc]: Right, removed them

**Review by efd6 — APPROVED** [review]

**Pull request merged**
